In [1]:
import os
import sys
import pprint
import pandas as pd

base_dir = '/media/tord/T7/Thesis_ssd/MasterThesis3.0'
os.chdir(base_dir)

from Classes.DataProcessing.LoadData import LoadData
from Classes.DataProcessing.DataHandler import DataHandler
from Classes.DataProcessing.DataGenerator import DataGenerator
from Classes.Modeling.GridSearchResultProcessor import GridSearchResultProcessor
from Classes.Modeling.CustomCallback import CustomCallback
from Classes.Scaling.ScalerFitter import ScalerFitter
from Classes.Scaling.MinMaxScalerFitter import MinMaxScalerFitter
from Classes.Scaling.StandardScalerFitter import StandardScalerFitter
from Classes.Modeling.ResultFitter import ResultFitter
import json

In [16]:
load_args = {
    'earth_explo_only' : True,
    'noise_earth_only' : False,
    'downsample' : True,
    'upsample' : True,
    'frac_diff' : 0.5,
    'seed' : 1,
    'subsample_size' : 0.25
}

loadData = LoadData(**load_args)

full_ds, train_ds, val_ds, test_ds = loadData.get_datasets()
noise_ds = loadData.noise_ds
handler = DataHandler(loadData)
dataGen = DataGenerator(loadData)
resultFitter = ResultFitter(loadData)

AttributeError: 'LoadData' object has no attribute 'getDatasets'

In [3]:
result_file_name = 'results_LSTM_NARROW_noiseNotNoise_timeAug_sscale_noiseAug_earlyS.csv'
model = resultFitter.fit_from_csv_and_index(result_file_name, 0, 2, 10, use_tensorboard = False, 
                               use_liveplots = True, use_custom_callback = False)

NameError: name 'resultFitter' is not defined

In [7]:
def f1_score(precision, recall):
    f1 = 2*((precision*recall)/(precision + recall))
    return f1

def create_f1_list(precision_df, recall_df):
    f1 = []
    for i in range(len(precision_df)):
        f1.append(f1_score(precision_df.loc[i], recall_df.loc[i]))
    return f1

#result_file_name = 'results_InceptionTime_noiseNotNoise_timeAug_sscale_noiseAug_earlyS.csv'
result_file_name = 'results_InceptionTime_NARROW_noiseNotNoise_timeAug_sscale_noiseAug_earlyS.csv'
df = GridSearchResultProcessor().get_results_df_by_name(result_file_name, 2)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

df_f1 = df.copy()
df_f1.columns=df_f1.columns.str.strip()
all_train_precision = df_f1['train_precision']
all_train_recall = df_f1['train_recall']
all_val_precision = df_f1['val_precision']
all_val_recall = df_f1['val_recall']
f1_train = create_f1_list(all_train_precision, all_train_recall)
f1_val = create_f1_list(all_val_precision, all_val_recall)
df_f1['train_f1'] = f1_train
df_f1['val_f1'] = f1_val
df_f1

batch_size  epochs  learning_rate optimizer  bottleneck_size  kernel_size  \
0          128     100        0.00001      adam               40           20   
1           64     100        0.00100      adam               26           20   
2          128     100        0.00100      adam               26           20   
3          512     100        0.00100      adam               26           20   
4          768     100        0.00100      adam               26           20   
5         1024     100        0.00100      adam               26           20   
6          256     100        0.00100      adam               26           20   
7          256     100        0.00100      adam               26           20   
8          256     100        0.00100      adam               26           20   
9          256     100        0.00100      adam               26           20   
10         256     100        0.00100      adam               26           20   
11         256     100        0.00100      adam               26           20   
12         256     100        0.00100      adam               26           20   
13         256     100        0.00100      adam               26           20   
14         256     100        0.00100      adam               26           20   
15         256     100        0.00100      adam               26           20   
16         256     100        0.00100      adam               22           20   
17         256     100        0.00100      adam               30           20   
18         256     100        0.00100      adam               26           20   
19         256     100        0.00100      adam               26           20   
20         256     100        0.00100      adam               26           20   
21         256     100        0.00100      adam               26           20   
22         256     100        0.00100      adam               26           20   
23         256     100        0.00100      adam               26           20   
24         256     100        0.10000      adam               26           20   
25         256     100        0.01000      adam               26           20   
26         256     100        0.00010      adam               26           20   
27         256     100        0.00001      adam               26           20   
28         256     100        0.00100      adam               26           20   
29         256     100        0.00100      adam               26           20   
30         256     100        0.00100      adam               26           20   
31         256     100        0.00100      adam               26           20   
32         256      80        0.00100      adam               26           20   
33         256      90        0.00100      adam               26           20   
34         256     110        0.00100      adam               26           20   
35         256     120        0.00100      adam               26           20   
36         256     100        0.00100      adam               26           20   
37         256     100        0.00100      adam               26           20   
38         256     100        0.00100      adam               26           20   
39         256     100        0.00100      adam               26           20   
40         256     100        0.00100      adam               26           20   
41         256     100        0.00100      adam               26           10   
42         256     100        0.00100      adam               26           30   
43         256     100        0.00100      adam               26           40   
44         256     100        0.00100      adam               26           50   
45         256     100        0.00100   rmsprop               26           20   
46         256     100        0.00100       sgd               26           20   

   module_activation module_output_activation  nr_modules  num_filters  \
0               relu                  sig

In [8]:
sorted_f1_df = df_f1.sort_values(by=['val_f1', 'train_f1'], axis = 0, ascending=False)
sorted_f1_df

batch_size  epochs  learning_rate optimizer  bottleneck_size  kernel_size  \
14         256     100        0.00100      adam               26           20   
15         256     100        0.00100      adam               26           20   
35         256     120        0.00100      adam               26           20   
38         256     100        0.00100      adam               26           20   
17         256     100        0.00100      adam               30           20   
43         256     100        0.00100      adam               26           40   
36         256     100        0.00100      adam               26           20   
18         256     100        0.00100      adam               26           20   
10         256     100        0.00100      adam               26           20   
39         256     100        0.00100      adam               26           20   
19         256     100        0.00100      adam               26           20   
22         256     100        0.00100      adam               26           20   
40         256     100        0.00100      adam               26           20   
13         256     100        0.00100      adam               26           20   
30         256     100        0.00100      adam               26           20   
23         256     100        0.00100      adam               26           20   
33         256      90        0.00100      adam               26           20   
26         256     100        0.00010      adam               26           20   
45         256     100        0.00100   rmsprop               26           20   
31         256     100        0.00100      adam               26           20   
9          256     100        0.00100      adam               26           20   
2          128     100        0.00100      adam               26           20   
34         256     110        0.00100      adam               26           20   
32         256      80        0.00100      adam               26           20   
37         256     100        0.00100      adam               26           20   
5         1024     100        0.00100      adam               26           20   
8          256     100        0.00100      adam               26           20   
42         256     100        0.00100      adam               26           30   
11         256     100        0.00100      adam               26           20   
41         256     100        0.00100      adam               26           10   
29         256     100        0.00100      adam               26           20   
16         256     100        0.00100      adam               22           20   
20         256     100        0.00100      adam               26           20   
25         256     100        0.01000      adam               26           20   
21         256     100        0.00100      adam               26           20   
4          768     100        0.00100      adam               26           20   
1           64     100        0.00100      adam               26           20   
44         256     100        0.00100      adam               26           50   
12         256     100        0.00100      adam               26           20   
3          512     100        0.00100      adam               26           20   
28         256     100        0.00100      adam               26           20   
27         256     100        0.00001      adam               26           20   
6          256     100        0.00100      adam               26           20   
7          256     100        0.00100      adam               26           20   
46         256     100        0.00100       sgd               26           20   
24         256     100        0.10000      adam               26           20   
0          128     100        0.00001      adam               40           20   

   module_activation module_output_activation  nr_modules  num_filters  \
14           sigmoid                     

In [9]:
sorted_f1_df = df_f1.sort_values(by=['val_precision', 'train_precision'], axis = 0, ascending=False)
sorted_f1_df

batch_size  epochs  learning_rate optimizer  bottleneck_size  kernel_size  \
35         256     120        0.00100      adam               26           20   
36         256     100        0.00100      adam               26           20   
31         256     100        0.00100      adam               26           20   
10         256     100        0.00100      adam               26           20   
17         256     100        0.00100      adam               30           20   
22         256     100        0.00100      adam               26           20   
13         256     100        0.00100      adam               26           20   
30         256     100        0.00100      adam               26           20   
2          128     100        0.00100      adam               26           20   
45         256     100        0.00100   rmsprop               26           20   
18         256     100        0.00100      adam               26           20   
11         256     100        0.00100      adam               26           20   
23         256     100        0.00100      adam               26           20   
34         256     110        0.00100      adam               26           20   
33         256      90        0.00100      adam               26           20   
9          256     100        0.00100      adam               26           20   
20         256     100        0.00100      adam               26           20   
19         256     100        0.00100      adam               26           20   
40         256     100        0.00100      adam               26           20   
32         256      80        0.00100      adam               26           20   
43         256     100        0.00100      adam               26           40   
39         256     100        0.00100      adam               26           20   
14         256     100        0.00100      adam               26           20   
42         256     100        0.00100      adam               26           30   
29         256     100        0.00100      adam               26           20   
37         256     100        0.00100      adam               26           20   
1           64     100        0.00100      adam               26           20   
41         256     100        0.00100      adam               26           10   
16         256     100        0.00100      adam               22           20   
8          256     100        0.00100      adam               26           20   
21         256     100        0.00100      adam               26           20   
38         256     100        0.00100      adam               26           20   
44         256     100        0.00100      adam               26           50   
15         256     100        0.00100      adam               26           20   
12         256     100        0.00100      adam               26           20   
26         256     100        0.00010      adam               26           20   
3          512     100        0.00100      adam               26           20   
4          768     100        0.00100      adam               26           20   
5         1024     100        0.00100      adam               26           20   
27         256     100        0.00001      adam               26           20   
25         256     100        0.01000      adam               26           20   
28         256     100        0.00100      adam               26           20   
24         256     100        0.10000      adam               26           20   
6          256     100        0.00100      adam               26           20   
7          256     100        0.00100      adam               26           20   
46         256     100        0.00100       sgd               26           20   
0          128     100        0.00001      adam               40           20   

   module_activation module_output_activation  nr_modules  num_filters  \
35           sigmoid                     

In [11]:
sorted_f1_df = df_f1.sort_values(by=['val_loss', 'train_loss'], axis = 0, ascending=True)
sorted_f1_df

batch_size  epochs  learning_rate optimizer  bottleneck_size  kernel_size  \
31         256     100        0.00100      adam               26           20   
35         256     120        0.00100      adam               26           20   
10         256     100        0.00100      adam               26           20   
13         256     100        0.00100      adam               26           20   
22         256     100        0.00100      adam               26           20   
1           64     100        0.00100      adam               26           20   
2          128     100        0.00100      adam               26           20   
17         256     100        0.00100      adam               30           20   
34         256     110        0.00100      adam               26           20   
30         256     100        0.00100      adam               26           20   
45         256     100        0.00100   rmsprop               26           20   
23         256     100        0.00100      adam               26           20   
20         256     100        0.00100      adam               26           20   
18         256     100        0.00100      adam               26           20   
9          256     100        0.00100      adam               26           20   
42         256     100        0.00100      adam               26           30   
11         256     100        0.00100      adam               26           20   
33         256      90        0.00100      adam               26           20   
39         256     100        0.00100      adam               26           20   
29         256     100        0.00100      adam               26           20   
32         256      80        0.00100      adam               26           20   
16         256     100        0.00100      adam               22           20   
40         256     100        0.00100      adam               26           20   
19         256     100        0.00100      adam               26           20   
8          256     100        0.00100      adam               26           20   
41         256     100        0.00100      adam               26           10   
21         256     100        0.00100      adam               26           20   
43         256     100        0.00100      adam               26           40   
44         256     100        0.00100      adam               26           50   
12         256     100        0.00100      adam               26           20   
3          512     100        0.00100      adam               26           20   
25         256     100        0.01000      adam               26           20   
4          768     100        0.00100      adam               26           20   
5         1024     100        0.00100      adam               26           20   
36         256     100        0.00100      adam               26           20   
28         256     100        0.00100      adam               26           20   
26         256     100        0.00010      adam               26           20   
37         256     100        0.00100      adam               26           20   
14         256     100        0.00100      adam               26           20   
38         256     100        0.00100      adam               26           20   
15         256     100        0.00100      adam               26           20   
24         256     100        0.10000      adam               26           20   
7          256     100        0.00100      adam               26           20   
46         256     100        0.00100       sgd               26           20   
6          256     100        0.00100      adam               26           20   
27         256     100        0.00001      adam               26           20   
0          128     100        0.00001      adam               40           20   

   module_activation module_output_activation  nr_modules  num_filters  \
31           sigmoid                     

In [12]:
df_f1.iloc[31]

batch_size                       256
epochs                           100
learning_rate                  0.001
optimizer                       adam
bottleneck_size                   26
kernel_size                       20
module_activation            sigmoid
module_output_activation        tanh
nr_modules                        12
num_filters                       40
output_activation            sigmoid
shortcut_activation          softmax
use_bottleneck                 False
use_residuals                   True
train_loss                  0.134154
train_accuracy              0.991666
train_precision              0.97313
train_recall                0.967562
val_loss                    0.326458
val_accuracy                 0.95694
val_precision               0.877855
val_recall                  0.854558
train_f1                    0.970338
val_f1                       0.86605
Name: 31, dtype: object

## Findings:
 - New optimal dropout_rate: 0.2 (train_acc: 98%, val_acc: 90%) Old 0.3 (train_acc: 96.8%, val_acc: 90.9%)
 - RELU and Sigmoid activation functions do not work with this model.
 - L1 regularization is counter productive (Less is more) so let l1 = 0.00
 - 32 start neurons appear to be optimal

In [161]:
df.iloc[40:]

batch_size  epochs  learning_rate  num_layers optimizer activation  \
40  64          33      0.20000        2           sgd       tanh        
41  64          33      0.10000        2           sgd       tanh        
42  64          33      0.00100        2           sgd       tanh        
43  64          33      0.00010        2           sgd       tanh        
44  64          33      0.00001        2           sgd       tanh        
45  64          33      0.01000        2           sgd       tanh        
46  64          33      0.01000        2           sgd       tanh        
47  64          33      0.01000        2           sgd       tanh        
48  64          33      0.01000        2           sgd       tanh        
49  64          33      0.01000        2           sgd       tanh        
50  64          33      0.01000        2           sgd       tanh        
51  64          33      0.01000        2           sgd       tanh        
52  64          33      0.01000        2           sgd       tanh        
53  64          33      0.01000        2           sgd       tanh        
54  32          33      0.01000        3           sgd       tanh        
55  32          33      0.01000        4           sgd       tanh        
56  32          33      0.01000        5           sgd       tanh        
57  32          33      0.01000        6           sgd       tanh        
58  24          33      0.01000        2           sgd       tanh        
59  28          33      0.01000        2           sgd       tanh        
60  34          33      0.01000        2           sgd       tanh        
61  36          33      0.01000        2           sgd       tanh        
62  32          33      0.10000        2           sgd       tanh        
63  32          33      0.05000        2           sgd       tanh        
64  32          33      0.02500        2           sgd       tanh        
65  32          33      0.00500        2           sgd       tanh        

         decay_sequence  dropout_rate  filters  kernel_size    l1_r   l2_r  \
40  [1, 2]               0.3           17       5            0.0001  0.001   
41  [1, 2]               0.3           17       5            0.0001  0.001   
42  [1, 2]               0.3           17       5            0.0001  0.001   
43  [1, 2]               0.3           17       5            0.0001  0.001   
44  [1, 2]               0.3           17       5            0.0001  0.001   
45  [1, 2]               0.3           17       5            0.0001  0.001   
46  [1, 2]               0.3           17       5            0.0001  0.001   
47  [1, 2]               0.3           17       5            0.0001  0.001   
48  [1, 2]               0.3           17       5            0.0001  0.001   
49  [1, 2]               0.3           17       5            0.0001  0.001   
50  [1, 2]               0.3           17       5            0.0001  0.001   
51  [1, 4]               0.3           17       5            0.0001  0.001   
52  [1, 0.5]             0.3           17       5            0.0001  0.001   
53  [1, 1]               0.3           17       5            0.0001  0.001   
54  [1, 2, 4]            0.3           17       5            0.0001  0.001   
55  [1, 2, 4, 6]         0.3           17       5            0.0001  0.001   
56  [1, 2, 4, 6, 8]      0.3           17       5            0.0001  0.001   
57  [1, 2, 4, 6, 8, 10]  0.3           17       5            0.0001  0.001   
58  [1, 2]               0.3           17       5            0.0001  0.001   
59  [1, 2]               0.3           17       5            0.0001  0.001   
60  [1, 2]               0.3           17       5            0.0001  0.001   
61  [1, 2]               0.3           17       5            0.0001  0.001   
62  [1, 2]               0.3           17       5            0.0001  0.001   
63  [1, 2]               0.3           17       5            0.0001  0.001   
64  [1, 2]               0.3           17       5      

In [162]:
def f1_score(precision, recall):
    f1 = 2*((precision*recall)/(precision + recall))
    return f1

def create_f1_list(precision_df, recall_df):
    f1 = []
    for i in range(len(precision_df)):
        f1.append(f1_score(precision_df.loc[i], recall_df.loc[i]))
    return f1

batch_size  epochs  learning_rate  num_layers optimizer activation  \
0   64          33      0.01000        2           sgd       tanh        
1   64          33      0.01000        2           sgd       tanh        
2   64          33      0.01000        2           sgd       tanh        
3   64          33      0.01000        2           sgd       tanh        
4   64          33      0.01000        2           sgd       tanh        
5   64          33      0.01000        2           sgd       tanh        
6   64          33      0.01000        2           sgd       tanh        
7   64          33      0.01000        2           sgd       tanh        
8   64          33      0.01000        2           adam      tanh        
9   64          33      0.01000        2           rmsprop   tanh        
10  64          33      0.01000        2           sgd       tanh        
11  64          33      0.01000        2           sgd       tanh        
12  64          33      0.01000        2           sgd       tanh        
13  64          33      0.01000        2           sgd       tanh        
14  64          33      0.01000        2           sgd       tanh        
15  64          33      0.01000        2           sgd       tanh        
16  64          33      0.01000        2           sgd       tanh        
17  64          33      0.01000        2           sgd       tanh        
18  64          29      0.01000        2           sgd       tanh        
19  64          37      0.01000        2           sgd       tanh        
20  64          33      0.01000        2           sgd       tanh        
21  64          33      0.01000        2           sgd       tanh        
22  64          33      0.01000        2           sgd       tanh        
23  64          33      0.01000        2           sgd       tanh        
24  64          33      0.01000        2           sgd       tanh        
25  64          33      0.01000        2           sgd       tanh        
26  64          33      0.01000        2           sgd       relu        
27  64          33      0.01000        2           sgd       sigmoid     
28  64          33      0.01000        2           sgd       softmax     
29  64          33      0.01000        2           sgd       tanh        
30  64          33      0.01000        2           sgd       tanh        
31  64          33      0.01000        1           sgd       tanh        
32  64          33      0.01000        3           sgd       tanh        
33  64          33      0.01000        4           sgd       tanh        
34  64          33      0.01000        5           sgd       tanh        
35  64          33      0.01000        6           sgd       tanh        
36  32          33      0.01000        2           sgd       tanh        
37  128         33      0.01000        2           sgd       tanh        
38  256         33      0.01000        2           sgd       tanh        
39  512         33      0.01000        2           sgd       tanh        
40  64          33      0.20000        2           sgd       tanh        
41  64          33      0.10000        2           sgd       tanh        
42  64          33      0.00100        2           sgd       tanh        
43  64          33      0.00010        2           sgd       tanh        
44  64          33      0.00001        2           sgd       tanh        
45  64          33      0.01000        2           sgd       tanh        
46  64          33      0.01000        2           sgd       tanh        
47  64          33      0.01000        2           sgd       tanh        
48  64          33      0.01000        2           sgd       tanh        
49  64          33      0.01000        2           sgd       tanh        
50  64          33      0.01000        2           sgd       tanh        
51  64          33      0.01000        2           sgd       tanh        
52  64          33      0.01000        2           sgd       tanh        
53  64  

In [170]:
sorted_f1_df

batch_size  epochs  learning_rate  num_layers optimizer activation  \
19  64          37      0.01000        2           sgd       tanh        
10  64          33      0.01000        2           sgd       tanh        
25  64          33      0.01000        2           sgd       tanh        
11  64          33      0.01000        2           sgd       tanh        
13  64          33      0.01000        2           sgd       tanh        
47  64          33      0.01000        2           sgd       tanh        
36  32          33      0.01000        2           sgd       tanh        
51  64          33      0.01000        2           sgd       tanh        
14  64          33      0.01000        2           sgd       tanh        
29  64          33      0.01000        2           sgd       tanh        
49  64          33      0.01000        2           sgd       tanh        
48  64          33      0.01000        2           sgd       tanh        
37  128         33      0.01000        2           sgd       tanh        
3   64          33      0.01000        2           sgd       tanh        
53  64          33      0.01000        2           sgd       tanh        
12  64          33      0.01000        2           sgd       tanh        
32  64          33      0.01000        3           sgd       tanh        
40  64          33      0.20000        2           sgd       tanh        
4   64          33      0.01000        2           sgd       tanh        
45  64          33      0.01000        2           sgd       tanh        
18  64          29      0.01000        2           sgd       tanh        
46  64          33      0.01000        2           sgd       tanh        
50  64          33      0.01000        2           sgd       tanh        
5   64          33      0.01000        2           sgd       tanh        
52  64          33      0.01000        2           sgd       tanh        
2   64          33      0.01000        2           sgd       tanh        
34  64          33      0.01000        5           sgd       tanh        
0   64          33      0.01000        2           sgd       tanh        
6   64          33      0.01000        2           sgd       tanh        
62  32          33      0.10000        2           sgd       tanh        
7   64          33      0.01000        2           sgd       tanh        
38  256         33      0.01000        2           sgd       tanh        
41  64          33      0.10000        2           sgd       tanh        
15  64          33      0.01000        2           sgd       tanh        
35  64          33      0.01000        6           sgd       tanh        
59  28          33      0.01000        2           sgd       tanh        
24  64          33      0.01000        2           sgd       tanh        
1   64          33      0.01000        2           sgd       tanh        
16  64          33      0.01000        2           sgd       tanh        
27  64          33      0.01000        2           sgd       sigmoid     
23  64          33      0.01000        2           sgd       tanh        
28  64          33      0.01000        2           sgd       softmax     
22  64          33      0.01000        2           sgd       tanh        
21  64          33      0.01000        2           sgd       tanh        
20  64          33      0.01000        2           sgd       tanh        
26  64          33      0.01000        2           sgd       relu        
61  36          33      0.01000        2           sgd       tanh        
17  64          33      0.01000        2           sgd       tanh        
60  34          33      0.01000        2           sgd       tanh        
39  512         33      0.01000        2           sgd       tanh        
42  64          33      0.00100        2           sgd       tanh        
9   64          33      0.01000        2           rmsprop   tanh        
33  64          33      0.01000        4           sgd       tanh        
54  32  

In [155]:
pd.to_numeric(df_f1['val_f1'], errors='coerce')

0     0.821132
1     0.804132
2     0.824117
3     0.832628
4     0.830515
5     0.828167
6     0.820769
7     0.818637
8     0.766957
9     0.785517
10    0.850172
11    0.843275
12    0.831845
13    0.840825
14    0.836419
15    0.809115
16    0.800293
17    0.790123
18    0.828972
19    0.856512
20    0.796238
21    0.796238
22    0.796238
23    0.796238
24    0.805288
25    0.845431
26    0.796238
27    0.796238
28    0.796238
29    0.836253
30   NaN      
31    0.761004
32    0.831517
33    0.784201
34    0.821750
35    0.806797
36    0.837878
37    0.833272
38    0.818396
39    0.789155
40    0.831131
41    0.815500
42    0.789035
43    0.583601
44    0.572993
45    0.829969
46    0.828561
47    0.838636
48    0.833555
49    0.835149
50    0.828306
51    0.836628
52    0.825959
53    0.832174
54    0.780612
55    0.775374
56    0.756599
57    0.712715
58    0.754318
59    0.806134
60    0.789657
61    0.795973
62    0.819888
63    0.761058
64   NaN      
Name: val_f1, dtype: floa

In [152]:
df_f1['val_f1'].isnull().values.any()

True

In [156]:
df_f1.columns

Index(['batch_size', 'epochs', 'learning_rate', 'num_layers', 'optimizer',
       'activation', 'decay_sequence', 'dropout_rate', 'filters',
       'kernel_size', 'l1_r', 'l2_r', 'output_layer_activation', 'padding',
       'start_neurons', 'train_loss', 'train_precision', 'train_accuracy',
       'train_recall', 'val_loss', 'val_precision', 'val_accuracy',
       'val_recall', 'train_f1', 'val_f1'],
      dtype='object')